### PR TITLE
Harden ls builtin: symlink safety, entry limits, pentest tests

### PR DESCRIPTION
**Pull request by matt-dz**
## Summary

- **Skip symlinks during `-R` recursion** to match GNU ls default behaviour and prevent symlink-loop DoS attacks. Uses `DirEntry.Type()` (does not follow symlinks) instead of relying on `FileInfo.IsDir()` after `e.Info()` (which follows symlinks).
- **Add `MaxDirEntries` (1,000) limit** per directory to prevent OOM from directories with millions of entries.
- **Paginated `ls`** with `--offset` and `--limit` flags for single-directory, non-recursive listings. Flags are silently ignored with `-R`.
- **Add pentest tests** covering symlink loops, symlink-to-directory follow prevention, recursion depth limits, entry count limits, flag injection, sandbox escape, path traversal, pagination, and filename edge cases.
- **Add scenario test** validating `ls -R` does not follow symlinks (asserted against bash).

## Intentional DoS tradeoffs

Three review threads are left **intentionally unresolved** because they flag behaviour that is a deliberate security tradeoff, not a bug:

1. **Hidden files consume `MaxDirEntries` quota** (`ls.go` readDir call) — The `MaxDirEntries` cap operates on raw directory reads, *before* hidden-file filtering. If we applied the cap after filtering, an attacker could create a directory with 999,999 dotfiles and 1 visible file — the shell would have to read all entries to find visible ones, defeating the DoS protection entirely. Code comment added in ce6c63e.

2. **`readDirLimited` truncates before global sort** (`allowed_paths.go`) — For directories larger than `maxRead`, the returned entries are sorted within the read window but may not be the globally-smallest names. Reading all entries to get globally-correct sorting would defeat the DoS protection — a directory with millions of files would OOM or stall. The truncation warning communicates that output is incomplete. Code comment added in ce6c63e.

## Test plan

- [x] `go test ./interp/builtins/ls/ -v -run TestLsPentest` — all pentest tests pass
- [x] `go test ./tests/ -v -run "TestShellScenarios/cmd/ls"` — all scenario tests pass
- [x] `go test ./interp/... ./tests/... -timeout 120s` — full test suite passes
- [x] `RSHELL_BASH_TEST=1 go test ./tests/ -run TestShellScenariosAgainstBash -timeout 120s` — bash comparison (requires Docker)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: interp/allowed_paths.go [hunk 1]
```diff
@@ -192,6 +192,61 @@ func (s *pathSandbox) readDir(ctx context.Context, path string) ([]fs.DirEntry,
...
+	const batchSize = 256
+	var entries []fs.DirEntry
+	truncated := false
+	var lastErr error
+	for {
+		batch, err := f.ReadDir(batchSize)
+		entries = append(entries, batch...)
+		if len(entries) > maxRead {
+			truncated = true
+			break
```
[the remark this answers is not in the document]
Comment (matt-dz) [reply interp/allowed_paths.go]: This is an intentional tradeoff. Reading all entries to get globally-sorted results would defeat the purpose of `readDirLimited` — a directory with millions of files would OOM or stall. The early break IS the DoS protection. The sort is correct within the read window (~`maxRead + 255` entries), and the truncation warning already communicates that output is incomplete. For directories within `MaxDirEntries` (1,000), all entries are read and sorting is globally correct.

**Comment by matt-dz** [comment]
@codex make a comprehensive code and security reviews
